core/src/main/java/com/yahoo/tagchowder/Parser.java

```diff
@@ -1185,6 +1204,26 @@ private String makeName(final char[] buff, final int offset, final int length) {
                 }
                 start = true;
                 dst.append(translateColons ? '_' : ch);
+            } else if (ampValidation) {
+                //if the beginning character starts with '[' or '{', ensure the brackets are balanced.
+                //allows [text], [[text]], {text}, {{text}}..etc
+                if (start && (ch == '[' || ch == '{'))  {
+                    if (areBracketsBalanced(buff, offset, length)) {
```

Comment:
The more I read this code, the more I feel it is not that efficient. Looks like you can going to scan the whole string whenever there is a '[' multiple times in following case. Is there any way to make it more efficient? 
[[[[[[[[[[[text]]]]]]]]]]
Maybe you can scan it once for areBracketsBalanced() and use that info for real scanning here?

Reply:
There should be one scan. The start flag changes to false.